File: deps/dates-calc/test/unit.ml
open Dates_calc.Dates

let date = Alcotest.testable format_date (fun x y -> compare_dates x y = 0)

let check_eqs eqs =
  List.iter
    (fun (d1, delta, d2) ->
      (* checks that d1 + delta = d2 under all rounding modes *)
      List.iter
        (fun rmode ->
          Alcotest.(check date)
            (Format.asprintf "%s + %3s = %s" d1 delta d2)
            (date_of_string d2)
            (add_dates ~round:rmode (date_of_string d1) (period_of_string delta)))
        [RoundUp; RoundDown; AbortOnRound])
    eqs

let check_extended_eqs eqs =
  List.iter
    (fun (d1, delta, dru, drd) ->
      (* checks that d1 +up delta = dru, d1+down delta = drd, d1+abort delta
         aborts withou AmbiguousComputation *)
      Alcotest.(check date)
        (Format.asprintf "%s +up %3s = %s" d1 delta dru)
        (date_of_string dru)
        (add_dates ~round:RoundUp (date_of_string d1) (period_of_string delta));
      Alcotest.(check date)
        (Format.asprintf "%s +down %3s = %s" d1 delta drd)
        (date_of_string drd)
        (add_dates ~round:RoundDown (date_of_string d1) (period_of_string delta));
      Alcotest.check_raises (Format.asprintf "%s +abort %3s raises" d1 delta)
        AmbiguousComputation (fun () ->
          let _ =
            add_dates ~round:AbortOnRound (date_of_string d1)
              (period_of_string delta)
          in
          ()))
    eqs

let read_csv ?(drop_header = true) filename =
  let ic = open_in filename in
  if drop_header then ignore(input_line ic);
  let rec loop acc =
    match input_line ic with
    | exception End_of_file -> List.rev acc
    | line ->
      loop ((String.split_on_char ';' line) :: acc)
  in
  loop []

let test_add_dates_exact () =
  let cases = read_csv "exact_computations.csv" in 
  let eqs = List.map
      (function
        | [d1; p; d2] -> (d1, p, d2)
        | l -> failwith (Format.asprintf "%d" (List.length l)))
      cases in
  check_eqs eqs

let test_add_dates_ambiguous () =
  let cases = read_csv "ambiguous_computations.csv" in
  let eqs = List.map
      (function
        | [d1; p; du; d_o] -> d1, p, du, d_o
        | _ -> assert false)
      cases in
  check_extended_eqs eqs

let test_first_last_day_of_month () =
  let cases = read_csv "first_last_day_of_month.csv" in
  let eqs = List.map
      (function
        | [d; df; dl] -> d, df, dl
        | _ -> assert false)
      cases in
  List.iter
    (fun (d_init, d_first, d_last) ->
       Alcotest.(check date)
         (Format.asprintf "first_day_of_month %s = %s" d_init d_first)
         (date_of_string d_first)
         (first_day_of_month @@ date_of_string d_init);
       Alcotest.(check date)
         (Format.asprintf "last_day_of_month %s = %s" d_init d_last)
         (date_of_string d_last)
         (last_day_of_month @@ date_of_string d_init)
    )
    eqs


(* Run it *)
let () =
  let open Alcotest in
  run "unit"
    [
      ( "add_dates",
        [
          test_case "exact" `Quick test_add_dates_exact;
          test_case "ambig" `Quick test_add_dates_ambiguous;
        ] );
      ( "first_last_day_of_month",
        [ test_case "all" `Quick test_first_last_day_of_month ] )
    ]

File: deps/dates-calc/test/unit.py
# Run from root directory of dates-calc with `python3 -m unittest test.unit`
import unittest
import csv
from lib_python.src.dates_calc import Date, Period, AbortOnRound, RoundUp, RoundDown, addup, adddown, AmbiguousComputation

class TestDates(unittest.TestCase):
    def read_csv(self, filename, drop_header=True):
        with open(filename) as csvfile:
            csvreader = csv.reader(csvfile, delimiter=';')
            r = []
            if drop_header: next(csvreader)
            for l in csvreader:
                tmp = [c.strip() for c in l]
                r.append(tmp)
            return r

    def test_add_dates_exact(self):
        cases = self.read_csv("test/exact_computations.csv") # unittest are supposed to be run from the root directory
        for c in cases:
            d = Date.from_string(c[0])
            p = Period.from_string(c[1])
            dr = Date.from_string(c[2])

            self.assertEqual(d + p, dr) # AbortOnRound
            self.assertEqual(d |addup| p, dr)
            self.assertEqual(d |adddown| p, dr)

    def test_add_dates_ambiguous(self):
        cases = self.read_csv("test/ambiguous_computations.csv") # unittest are supposed to be run from the root directory
        for c in cases:
            d = Date.from_string(c[0])
            p = Period.from_string(c[1])
            dup = Date.from_string(c[2])
            ddo = Date.from_string(c[3])

            with self.assertRaises(AmbiguousComputation):
                d + p
            self.assertEqual(d |addup| p, dup)
            self.assertEqual(d |adddown| p, ddo)

    def test_first_last_day_of_month(self):
        cases = self.read_csv("test/first_last_day_of_month.csv")
        for c in cases:
            d = Date.from_string(c[0])
            df = Date.from_string(c[1])
            dl = Date.from_string(c[2])

            d_f = d.first_day_of_month()
            d_l = d.last_day_of_month()

            self.assertEqual(df, d_f)
            self.assertEqual(dl, d_l)

if __name__ == "__main__":
    unittest.main(verbosity)
